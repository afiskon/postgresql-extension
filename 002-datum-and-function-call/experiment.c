#include <postgres.h>
#include <utils/builtins.h>
#include <datatype/timestamp.h>
#include <utils/timestamp.h>

PG_MODULE_MAGIC;
PG_FUNCTION_INFO_V1(timestamp_in_out_test);
PG_FUNCTION_INFO_V1(timestamptz_zone_test);

Datum
timestamp_in_out_test(PG_FUNCTION_ARGS)
{
    const char* tstamp_str = "Jan 01 00:00:00 2010";
    Timestamp tstamp;

    tstamp = DatumGetTimestamp(DirectFunctionCall3(timestamp_in,
        CStringGetDatum(tstamp_str),
        ObjectIdGetDatum(InvalidOid),
        Int32GetDatum(-1)));

    tstamp_str = DatumGetCString(DirectFunctionCall1(timestamp_out,
        TimestampGetDatum(tstamp)));

    PG_RETURN_TEXT_P(CStringGetTextDatum(tstamp_str));
}

Datum
timestamptz_zone_test(PG_FUNCTION_ARGS)
{
    Timestamp result;
    TimestampTz timestamptz = TimestampTzGetDatum(PG_GETARG_DATUM(0));
    const char* tzname = "Europe/Moscow";

    if(TIMESTAMP_NOT_FINITE(timestamptz)) {
        // Timestamp and TimestampTz internally are the same in PostgreSQL
        // The TZ-version of TIMESTAMP_NOT_FINITE() is not even defined!
        result = (Timestamp)timestamptz;
        PG_RETURN_TIMESTAMP(result);
    }

    if(PG_NARGS() > 1) {
        tzname = TextDatumGetCString(PG_GETARG_DATUM(1));
    }

    // The code is equal to 'timestamptz AT TIME ZONE tzname'
    result = DatumGetTimestamp(DirectFunctionCall2(timestamptz_zone,
            CStringGetTextDatum(tzname),
            TimestampTzGetDatum(timestamptz)));

    PG_RETURN_TIMESTAMP(result);
}
