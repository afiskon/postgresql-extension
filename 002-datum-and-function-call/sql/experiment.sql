CREATE EXTENSION experiment;

SELECT timestamp_in_out_test();
SELECT timestamptz_zone_test(null);
SELECT timestamptz_zone_test('-infinity');
SELECT timestamptz_zone_test('2010-01-01 00:00:00 MSK');
SELECT timestamptz_zone_test('2010-01-01 00:00:00 HST', 'Pacific/Honolulu');

\set ON_ERROR_STOP 0
SELECT timestamptz_zone_test('2010-01-01 00:00:00 GMT', 'Europe/Ololondon');
\set ON_ERROR_STOP 1

DROP EXTENSION experiment;