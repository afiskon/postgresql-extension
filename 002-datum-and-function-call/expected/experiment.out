CREATE EXTENSION experiment;
SELECT timestamp_in_out_test();
  timestamp_in_out_test   
--------------------------
 Fri Jan 01 00:00:00 2010
(1 row)

SELECT timestamptz_zone_test(null);
 timestamptz_zone_test 
-----------------------
 
(1 row)

SELECT timestamptz_zone_test('-infinity');
 timestamptz_zone_test 
-----------------------
 -infinity
(1 row)

SELECT timestamptz_zone_test('2010-01-01 00:00:00 MSK');
  timestamptz_zone_test   
--------------------------
 Fri Jan 01 00:00:00 2010
(1 row)

SELECT timestamptz_zone_test('2010-01-01 00:00:00 HST', 'Pacific/Honolulu');
  timestamptz_zone_test   
--------------------------
 Fri Jan 01 00:00:00 2010
(1 row)

\set ON_ERROR_STOP 0
SELECT timestamptz_zone_test('2010-01-01 00:00:00 GMT', 'Europe/Ololondon');
ERROR:  time zone "Europe/Ololondon" not recognized
\set ON_ERROR_STOP 1
DROP EXTENSION experiment;
